Add support for Draft flag on GitHub

hluk opened the pull request.

hluk committed:
Add support for Draft flag on GitHub

lcarva commented on reviewrot/githubstack.py:
```diff
@@ -159,9 +159,13 @@ def get_reviews(self, uname, repo_name, age=None, show_last_comment=None):
 
+            title = pr.title
+            if pr.draft:
+                title = f"Draft: {title}"
```
Could use the `WIP:` prefix instead so the UI also classifies it correctly. Maybe something like this:
```python
title = pr.title
if pr.draft and not title.upper().startswith("WIP"):
    title = "WIP: " + title
```

hluk replied:
Good point, updated. "Draft:" would also work I think but it is OK to use perhaps more common "WIP".

lcarva replied:
Just `WIP`: https://github.com/redhat-aqe/review-rot/blob/3436ce9b9fd4d6aaba3fe4f4dc707cc040b85f56/web/js/site.js

hluk replied:
Oh, I see, I forgot about the UI part.

hluk commented:
@pbortlov Please review.

hluk committed:
Fix argument name

hluk committed:
Re-format according to black

hluk committed:
Use project or group avatar for GitLab repos

hluk committed:
Add workaround for Gerrit error "Multiple changes found"

hluk committed:
Fix iterating all GitLab comments

Fixes the following warning:

    reviewrot/gitlabstack.py:227: UserWarning: Calling a `list()` method
    without specifying `get_all=True` or `iterator=True` will return a
    maximum of 20 items.

hluk committed:
Fix requirements.txt

parasense commented on test/github_tests/test_github.py:
```diff
@@ -370,3 +370,50 @@ def test_request_reviews_without_repo(self, mock_get_reviews, mock_github_patch)
+        mock_check_request_state.return_value = True
+
+        for title, expected_title in [
+            ("dummy_title", "WIP: dummy_title"),
+            ("WIP: dummy_title", "WIP: dummy_title"),
+            ("Draft: dummy_title", "Draft: dummy_title"),
```
Might also consider validating all-lower case.
The regex seems to allow mixed case by `re.IGNORECASE`.

parasense replied:
Might also try testing the `[WIP] `  square-brace variation.

hluk replied:
I have not changed this, only reformatted the code with black.

hluk replied:
Ah, actually, my mistake, that is the oldest commit in this PR. :facepalm:

parasense commented on web/js/site.js:
```diff
@@ -17,7 +17,7 @@ $(document).ready(function() {
 
 		// A user needs to change this value to get their site to work.
-		url: 'https://raw.githubusercontent.com/nirzari/review-rot/master/web/js/default-data.json',
+		url: 'https://exd.pages.redhat.com/lholecek/rhelwf-reviews/web/js/default-data.json',
```
This doesn't work for me.
I takes me to an internal (RH) gitlab login page.

hluk replied:
I added this change accidentally. I will revert it.

parasense commented on tox.ini:
```diff
@@ -36,6 +37,12 @@ commands =
         --diff \
         reviewrot/ test/
 
+[testenv:black-format]
+skip_install = true
+deps =
+    black
+commands =
+    black reviewrot/ test/
```
So for context, this project previously didn't have any `flake8` or `black` linting. 
I wrote a previous patch to PEP8 bomb the project, and added these linters to Tox.

My concern here is how this `black-format` would auto format, and I'm not 100% sure that's the best approach. 

I'd rather see folks use Tox locally, prior to pushing, and amend their commits if necessary. That way all code changes are tracked by a commit end-to-end, and not modified in-flight. Besides that, I also wonder what happens when folks GPG sign commits that get modified by a workflow, does the attestation get lost in translation?

hluk replied:
`tox -e black-format` is only meant to be run manually (not triggered with `tox`). It is just a shortcut to the specific `black` command so you do not need to remember it (or install the utility).

`tox -e black` should run automatically and fail if reformat is needed (we use this approach in other internal repos too).

parasense commented on reviewrot/utils.py:
```diff
@@ -0,0 +1,19 @@
+
+WIP_REGEX_PATTERN = r"^(\[WIP\]\s*|WIP:\s*|WIP\s+|Draft:)+\s*"
+
+
```
I played around with the regex here:
https://regex101.com/r/m6YF8I/1

hluk replied:
Changed. Thanks. :+1:
